src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/BannedTypesAnalyzer.cs

```diff
@@ -0,0 +1,106 @@
+        {
+            SymbolKind.Method,
+            SymbolKind.Field,
+            SymbolKind.Property,
+            SymbolKind.Parameter,
+            SymbolKind.Event,
+            SymbolKind.NamedType,
+            SymbolKind.Local,
+        };
+
+        public override void Analyze(ISymbolAnalysisContext context)
+        {
+            Debug.WriteLine($"{context.Symbol}");
+
+            if (IsAzureCore(context.Symbol.ContainingAssembly))
+            {
+                return;
+            }
+
+            switch (context.Symbol)
+            {
+                case IParameterSymbol parameterSymbol:
+                    CheckType(context, parameterSymbol.Type, parameterSymbol);
+                    break;
+                case IMethodSymbol methodSymbol:
+                    CheckType(context, methodSymbol.ReturnType, methodSymbol);
+                    break;
+                case IEventSymbol eventSymbol:
+                    CheckType(context, eventSymbol.Type, eventSymbol);
+                    break;
+                case IPropertySymbol propertySymbol:
+                    CheckType(context, propertySymbol.Type, propertySymbol);
+                    break;
+                case IFieldSymbol fieldSymbol:
+                    CheckType(context, fieldSymbol.Type, fieldSymbol);
+                    break;
+                case ILocalSymbol localSymbol:
+                    CheckType(context, localSymbol.Type, localSymbol);
+                    break;
+                case INamedTypeSymbol namedTypeSymbol:
+                    CheckType(context, namedTypeSymbol.BaseType, namedTypeSymbol);
+                    foreach (var iface in namedTypeSymbol.Interfaces)
+                    {
+                        CheckType(context, iface, namedTypeSymbol);
+                    }
+                    break;
+            }
+
+            Debug.WriteLine($"done");
+        }
+
+        private static void CheckType(ISymbolAnalysisContext context, ITypeSymbol type, ISymbol symbol)
+        {
+            if (type is INamedTypeSymbol namedTypeSymbol)
+            {
+                if (IsBannedType(namedTypeSymbol))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0020, symbol.Locations.First(), BannedTypesMessageArgs), symbol);
+                }
+
+                if (namedTypeSymbol.IsGenericType)
+                {
+                    foreach (var typeArgument in namedTypeSymbol.TypeArguments)
+                    {
+                        CheckType(context, typeArgument, symbol);
+                    }
+                }
+            }
+        }
+
+        private static bool IsBannedType(INamedTypeSymbol namedTypeSymbol)
+        {
+            return BannedTypes.Contains($"{namedTypeSymbol.ContainingNamespace}.{namedTypeSymbol.Name}");
+        }
+
+        private static bool IsAzureCore(IAssemblySymbol assembly)
+        {
+            return assembly.Name.Equals("Azure.Core");
```

Comment:
Should we allow anything in the Azure.Core family - such as Azure.Core.Experimental?

Reply:
Good point!  Added it.